Lib/pickle.py

```diff
@@ -216,6 +216,22 @@ def write(self, data):
         else:
             return self.file_write(data)
 
+    def write_large_bytes(self, header, payload):
+        write = self.file_write
+        if self.current_frame:
+            # Terminate the current frame to write the next frame directly
+            # into the underlying file to skip the unnecessary memory
+            # allocations of a large temporary buffer.
+            self.commit_frame(force=True)
+            frame_size = len(header) + len(payload)
+            write(FRAME + pack("<Q", frame_size))
```

Comment:
Is it worth to create a new frame containing a single bytes/string object? This will increase the size of the file, but will not decrease the number of file reading operations (opcode + frame/bytes/string size + payload).

Reply:
I think the presence of a frame is required by protocol 4. The overhead is likely to be small because it's only for large enough bytes instances.

Reply:
[PEP 3154](https://www.python.org/dev/peps/pep-3154/#framing) is not very explicit whether framing is required or optional. The existing implementations (Python and C) would never dump a frameless opcode as far as I understand so it might be safer to follow that behavior in the new nocopy code path but I can change that if you think this is being too conservative.

Reply:
I remember that some people voiced concerns about the frame header overhead (however small it is), so I worded the PEP so that it isn't made mandatory (and the unpickler should be able to handle both with and without frames). But I do think the standard pickler implementations should always emit frames in protocol 4.

Reply:
Frames are optional.

If don't create a frame for a separate bytes object, this could decrease memory usage on unpickling. You could directly use the result of the `read()` method without creating a buffer.

Reply:
Indeed I observed at least one useless memory copy at load time for large objects. I have not investigated yet, however, it's there both with the frameless protocol 3 and with protocol 4 with framing enabled. Anyway, it would be best to choose a design that allows for no-copy semantics of large objects both at dump and load times.